pkg/proxy/conntrack/conntrack.go

```diff
@@ -21,123 +21,54 @@ package conntrack
-func (ct *execCT) ClearEntriesForIP(ip string, protocol v1.Protocol) error {
-	parameters := parametersWithFamily(utilnet.IsIPv6String(ip), "-D", "--orig-dst", ip, "-p", protoStr(protocol))
-	err := ct.exec(parameters...)
-	if err != nil && !strings.Contains(err.Error(), noConnectionToDelete) {
-		// TODO: Better handling for deletion failure. When failure occur, stale udp connection may not get flushed.
-		// These stale udp connection will keep black hole traffic. Making this a best effort operation for now, since it
-		// is expensive to baby-sit all udp connections to kubernetes services.
-		return fmt.Errorf("error deleting connection tracking state for UDP service IP: %s, error: %v", ip, err)
-	}
-	return nil
-}
-
-// ClearEntriesForPort is part of Interface
-func (ct *execCT) ClearEntriesForPort(port int, isIPv6 bool, protocol v1.Protocol) error {
-	if port <= 0 {
-		return fmt.Errorf("wrong port number. The port number must be greater than zero")
-	}
-	parameters := parametersWithFamily(isIPv6, "-D", "-p", protoStr(protocol), "--dport", strconv.Itoa(port))
-	err := ct.exec(parameters...)
-	if err != nil && !strings.Contains(err.Error(), noConnectionToDelete) {
-		return fmt.Errorf("error deleting conntrack entries for UDP port: %d, error: %v", port, err)
-	}
-	return nil
-}
+	netlinkFilters := make([]netlink.CustomConntrackFilter, 0, len(filters))
+	netlinkFilters = append(netlinkFilters, filters...)
 
-// ClearEntriesForNAT is part of Interface
-func (ct *execCT) ClearEntriesForNAT(origin, dest string, protocol v1.Protocol) error {
-	parameters := parametersWithFamily(utilnet.IsIPv6String(origin), "-D", "--orig-dst", origin, "--dst-nat", dest,
-		"-p", protoStr(protocol))
-	err := ct.exec(parameters...)
-	if err != nil && !strings.Contains(err.Error(), noConnectionToDelete) {
-		// TODO: Better handling for deletion failure. When failure occur, stale udp connection may not get flushed.
-		// These stale udp connection will keep black hole traffic. Making this a best effort operation for now, since it
-		// is expensive to baby sit all udp connections to kubernetes services.
-		return fmt.Errorf("error deleting conntrack entries for UDP peer {%s, %s}, error: %v", origin, dest, err)
-	}
-	return nil
-}
-
-// ClearEntriesForPortNAT is part of Interface
-func (ct *execCT) ClearEntriesForPortNAT(dest string, port int, protocol v1.Protocol) error {
-	if port <= 0 {
-		return fmt.Errorf("wrong port number. The port number must be greater than zero")
-	}
-	parameters := parametersWithFamily(utilnet.IsIPv6String(dest), "-D", "-p", protoStr(protocol), "--dport", strconv.Itoa(port), "--dst-nat", dest)
-	err := ct.exec(parameters...)
-	if err != nil && !strings.Contains(err.Error(), noConnectionToDelete) {
-		return fmt.Errorf("error deleting conntrack entries for UDP port: %d, error: %v", port, err)
+	n, err := ct.handler.ConntrackDeleteFilters(netlink.ConntrackTable, netlink.InetFamily(ipFamily), netlinkFilters...)
```

Comment:
I really don't want to miss the logging of the deleted entries, this is my suggestion
```
	if klog.V(5).Enabled() {
                table, err := netlinkg.ConntrackTableList()
                 ...
                 defer func() {
                         newTable, err := netlinkg.ConntrackTableList()
		         klog.V(logLevel).Infof("deleted conntrack entries: %v", diff(table, newTable)
                 }()
	}
```

Reply:
The entries which have been deleted will be still logged in this case.
We are logging the flow which is filtered for deleting in **MatchConntrackFlow**, vishw/netlink will call this method in for all the filters in a loop, before requesting the flow deletion flow entry will be logged.

https://github.com/vishvananda/netlink/blob/v1.3.0/conntrack_linux.go#L164

Reply:
yeah, but you don;t know if it was actually deleted or not ... for debugging this issue is import to know what entries were actually deleted  see https://github.com/kubernetes/kubernetes/issues/126130#issuecomment-2291174088

> 
> I0815 12:15:14.624545       1 cleanup.go:63] "Deleting conntrack stale entries for services" IPs=[]
> I0815 12:15:14.624580       1 cleanup.go:69] "Deleting conntrack stale entries for services" nodePorts=[]
> I0815 12:15:14.624646       1 conntrack.go:66] "Clearing conntrack entries" parameters=["-D","--orig-dst","10.12.201.0","--dst-nat","10.11.4.22","-p","udp"]
> I0815 12:15:14.637712       1 conntrack.go:71] "Conntrack entries deleted" output=<
> 	udp      17 23 src=10.11.4.30 dst=10.12.201.0 sport=34152 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.30 sport=8125 dport=34152 mark=0 use=1
> 	udp      17 26 src=10.11.4.38 dst=10.12.201.0 sport=50751 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.38 sport=8125 dport=50751 mark=0 use=1
> 	udp      17 28 src=10.11.4.5 dst=10.12.201.0 sport=48931 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.5 sport=8125 dport=48931 mark=0 use=1
> 	udp      17 28 src=10.11.4.5 dst=10.12.201.0 sport=42608 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.5 sport=8125 dport=42608 mark=0 use=1
> 	udp      17 21 src=10.11.4.56 dst=10.12.201.0 sport=50648 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.56 sport=8125 dport=50648 mark=0 use=1
> 	udp      17 21 src=10.11.4.64 dst=10.12.201.0 sport=60746 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.64 sport=8125 dport=60746 mark=0 use=1
> 	udp      17 18 src=10.11.4.20 dst=10.12.201.0 sport=40877 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.20 sport=8125 dport=40877 mark=0 use=1
> 	udp      17 20 src=10.11.4.55 dst=10.12.201.0 sport=56370 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.55 sport=8125 dport=56370 mark=0 use=1
> 	udp      17 12 src=10.11.4.64 dst=10.12.201.0 sport=37004 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.64 sport=8125 dport=37004 mark=0 use=1
> 	udp      17 28 src=10.11.4.50 dst=10.12.201.0 sport=57899 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.50 sport=8125 dport=57899 mark=0 use=1
> 	udp      17 29 src=10.11.4.113 dst=10.12.201.0 sport=58278 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.113 sport=8125 dport=58278 mark=0 use=1
> 	udp      17 26 src=10.11.4.8 dst=10.12.201.0 sport=40499 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.8 sport=8125 dport=40499 mark=0 use=1
> 	udp      17 20 src=10.11.4.67 dst=10.12.201.0 sport=48386 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.67 sport=8125 dport=48386 mark=0 use=1
> 	udp      17 29 src=10.11.4.50 dst=10.12.201.0 sport=55565 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.50 sport=8125 dport=55565 mark=0 use=1
> 	udp      17 29 src=10.11.4.61 dst=10.12.201.0 sport=36647 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.61 sport=8125 dport=36647 mark=0 use=1
> 	udp      17 29 src=10.11.4.59 dst=10.12.201.0 sport=52344 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.59 sport=8125 dport=52344 mark=0 use=1
> 	udp      17 28 src=10.11.4.5 dst=10.12.201.0 sport=35652 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.5 sport=8125 dport=35652 mark=0 use=1
> 	udp      17 22 src=10.11.4.145 dst=10.12.201.0 sport=46692 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.145 sport=8125 dport=46692 mark=0 use=1
> 	udp      17 28 src=10.11.4.5 dst=10.12.201.0 sport=43098 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.5 sport=8125 dport=43098 mark=0 use=1
> 	udp      17 29 src=10.11.4.59 dst=10.12.201.0 sport=51314 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.59 sport=8125 dport=51314 mark=0 use=1
> 	udp      17 29 src=10.11.4.70 dst=10.12.201.0 sport=48682 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.70 sport=8125 dport=48682 mark=0 use=1
> 	udp      17 20 src=10.11.4.65 dst=10.12.201.0 sport=50380 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.65 sport=8125 dport=50380 mark=0 use=1
> 	udp      17 16 src=10.11.4.217 dst=10.12.201.0 sport=44527 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.217 sport=8125 dport=44527 mark=0 use=1
> 	udp      17 21 src=10.11.4.66 dst=10.12.201.0 sport=37202 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.66 sport=8125 dport=37202 mark=0 use=1
> 	udp      17 24 src=10.11.4.31 dst=10.12.201.0 sport=38039 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.31 sport=8125 dport=38039 mark=0 use=1
> 	udp      17 22 src=10.11.4.69 dst=10.12.201.0 sport=39822 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.69 sport=8125 dport=39822 mark=0 use=1
> 	udp      17 29 src=10.11.4.143 dst=10.12.201.0 sport=52499 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.143 sport=8125 dport=52499 mark=0 use=1
> 	udp      17 21 src=10.11.4.64 dst=10.12.201.0 sport=46370 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.64 sport=8125 dport=4conntrack v1.4.7 (conntrack-tools): 37 flow entries have been deleted.
> 	6370 mark=0 use=1
> 	udp      17 1 src=10.11.4.27 dst=10.12.201.0 sport=41082 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.27 sport=8125 dport=41082 mark=0 use=1
> 	udp      17 28 src=10.11.4.5 dst=10.12.201.0 sport=40601 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.5 sport=8125 dport=40601 mark=0 use=1
> 	udp      17 24 src=10.11.4.64 dst=10.12.201.0 sport=57893 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.64 sport=8125 dport=57893 mark=0 use=1
> 	udp      17 27 src=10.11.4.2 dst=10.12.201.0 sport=32932 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.2 sport=8125 dport=32932 mark=0 use=1
> 	udp      17 16 src=10.11.4.64 dst=10.12.201.0 sport=47022 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.64 sport=8125 dport=47022 mark=0 use=1
> 	udp      17 28 src=10.11.4.5 dst=10.12.201.0 sport=40394 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.5 sport=8125 dport=40394 mark=0 use=1
> 	udp      17 29 src=10.11.4.61 dst=10.12.201.0 sport=47876 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.61 sport=8125 dport=47876 mark=0 use=1
> 	udp      17 29 src=10.11.4.24 dst=10.12.201.0 sport=40378 dport=8125 [UNREPLIED] src=10.11.4.22 dst=10.11.4.24 sport=8125 dport=40378 mark=0 use=1
> 	udp      17 30 src=10.11.4.5 dst=10.12.201.0 sport=46634 dport=8125 [UNREPLIED] src=10.11.4.22 ds
>

Reply:
this can be resolved